stored_responses/stored_responses.go

```diff
@@ -75,53 +74,62 @@ func extractStoredResponsesIds(impInfo []ImpExtPrebidData,
+		if impExtPrebid.StoredAuctionResponse != nil {
+			if len(impExtPrebid.StoredAuctionResponse.ID) == 0 {
 				return nil, nil, nil, nil, fmt.Errorf("request.imp[%d] has ext.prebid.storedauctionresponse specified, but \"id\" field is missing ", index)
 			}
-			allStoredResponseIDs = append(allStoredResponseIDs, impData.ImpExtPrebid.StoredAuctionResponse.ID)
+			allStoredResponseIDs = append(allStoredResponseIDs, impExtPrebid.StoredAuctionResponse.ID)
 
-			impAuctionResponseIDs[impId] = impData.ImpExtPrebid.StoredAuctionResponse.ID
+			impAuctionResponseIDs[impId] = impExtPrebid.StoredAuctionResponse.ID
 
 		}
-		if len(impData.ImpExtPrebid.StoredBidResponse) > 0 {
+		if len(impExtPrebid.StoredBidResponse) > 0 {
+
+			// bidders can be specified in imp.ext and in imp.ext.prebid.bidders
+			allBidderNames := make([]string, 0)
+			for bidderName := range impExtPrebid.Bidder {
+				allBidderNames = append(allBidderNames, bidderName)
+			}
+			for extData := range impExt.GetExt() {
+				// no bidders will not be processed
+				allBidderNames = append(allBidderNames, extData)
+			}
 
 			bidderStoredRespId := make(map[string]string)
 			bidderReplaceImpId := make(map[string]bool)
-			for _, bidderResp := range impData.ImpExtPrebid.StoredBidResponse {
+			for _, bidderResp := range impExtPrebid.StoredBidResponse {
 				if len(bidderResp.ID) == 0 || len(bidderResp.Bidder) == 0 {
 					return nil, nil, nil, nil, fmt.Errorf("request.imp[%d] has ext.prebid.storedbidresponse specified, but \"id\" or/and \"bidder\" fields are missing ", index)
 				}
-				bidderName := bidderResp.Bidder
-				normalizedCoreBidder, ok := openrtb_ext.NormalizeBidderName(bidderResp.Bidder)
-				if ok {
-					bidderName = normalizedCoreBidder.String()
-				}
-				//check if bidder is valid/exists
-				if _, isValid := bidderMap[bidderName]; !isValid {
-					return nil, nil, nil, nil, fmt.Errorf("request.imp[impId: %s].ext.prebid.bidder contains unknown bidder: %s. Did you forget an alias in request.ext.prebid.aliases?", impId, bidderResp.Bidder)
-				}
-				// bidder is unique per one bid stored response
-				// if more than one bidder specified the last defined bidder id will take precedence
-				bidderStoredRespId[bidderResp.Bidder] = bidderResp.ID
-				impBiddersWithBidResponseIDs[impId] = bidderStoredRespId
-
-				// stored response config can specify if imp id should be replaced with imp id from request
-				replaceImpId := true
-				if bidderResp.ReplaceImpId != nil {
-					// replaceimpid is true if not specified
-					replaceImpId = *bidderResp.ReplaceImpId
-				}
-				bidderReplaceImpId[bidderResp.Bidder] = replaceImpId
-				impBidderReplaceImp[impId] = bidderReplaceImpId
 
-				//storedAuctionResponseIds are not unique, but fetch will return single data for repeated ids
-				allStoredResponseIDs = append(allStoredResponseIDs, bidderResp.ID)
+				for _, bidderName := range allBidderNames {
+					if strings.EqualFold(bidderName, bidderResp.Bidder) {
```

Comment:
Should we add a `break` at the end of this `if` block so that the first stored response is chosen if multiple stored responses are specified for the same bidder?

Reply:
No, because we need to build a map of bidderName to stored response. Same stored response should be applied to any "Appnexus" or "APPnexus" bidder in imp.

Reply:
Gotcha, can we make line 116 something like: `if _, found := bidderStoredRespId[bidderName]; !found && strings.EqualFold(bidderName, bidderResp.Bidder) {` to ensure that the first stored response is used instead of the last if there are multiple for the same bidder?

Reply:
Yes! Added. And a unit test.